File: leetcode/00678_valid_parenthesis_string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        def is_balanced(it, add: str) -> bool:
            balance = 0
            for i in it:
                balance += 1 if s[i] in add else -1
                if balance < 0:
                    return False
            return True
        n = len(s)    
        return is_balanced(range(n), '(*') and is_balanced(range(n - 1, -1, -1), '*)')

File: leetcode/00678_valid_parenthesis_string.rs
impl Solution {
    pub fn check_valid_string(s: String) -> bool {
        let mut low = 0;
        let mut high = 0;
        for c in s.chars() {
            low += 1 - 2 * (c != '(') as i8;
            high += 1 - 2 * (c == ')') as i8;
            if high < 0 { return false; }
            low = low.max(0);
        }
        low == 0
    }
}

